src/store/RelayContext.js

```diff
@@ -0,0 +1,269 @@
+   * will contain exactly one element.
+   */
+  readQuery(
+    root: RelayQuery.Root,
+    options?: StoreReaderOptions
+  ): Array<?StoreReaderData> {
+    const storageKey = root.getStorageKey();
+    const results = [];
+    forEachRootCallArg(root, identifyingArgValue => {
+      let data;
+      const dataID = this._storeData.getQueuedStore()
+        .getDataID(storageKey, identifyingArgValue);
+      if (dataID != null) {
+        data = this.read(root, dataID, options);
+      }
+      results.push(data);
+    });
+    return results;
+  }
+
+  /**
+   * Reads and subscribes to query data anchored at the supplied data ID. The
+   * returned observable emits updates as the data changes over time.
+   */
+  observe(
+    fragment: RelayQuery.Fragment,
+    dataID: DataID
+  ): Observable<?StoreReaderData> {
+    const fragmentPointer = new GraphQLFragmentPointer(
+      fragment.isPlural()? [dataID] : dataID,
+      fragment
+    );
+    return new RelayQueryResultObservable(this._storeData, fragmentPointer);
+  }
+
+  applyUpdate(
+    mutation: RelayMutation,
+    callbacks?: RelayMutationTransactionCommitCallbacks
+  ): RelayMutationTransaction {
+    return this._storeData.getMutationQueue().createTransaction(
+      mutation,
+      callbacks
+    );
+  }
+
+  update(
+    mutation: RelayMutation,
+    callbacks?: RelayMutationTransactionCommitCallbacks
+  ): void {
+    this.applyUpdate(mutation, callbacks).commit();
+  }
+
+  /**
+   * Initializes garbage collection: must be called before any records are
+   * fetched. When records are collected after calls to `scheduleCollection` or
+   * `scheduleCollectionFromNode`, records are collected in steps, with a
+   * maximum of `stepLength` records traversed in a step. Steps are scheduled
+   * via `RelayTaskScheduler`.
+   */
+  initializeGarbageCollection(stepLength: number): void {
+    invariant(
+        stepLength > 0,
+        'RelayGarbageCollection: step length must be greater than zero, got ' +
+        '`%s`.',
+        stepLength
+    );
+    this._storeData.initializeGarbageCollector(scheduler);
+
+    const pendingQueryTracker = this._storeData.getPendingQueryTracker();
+
+    function scheduler(run: () => boolean): void {
+      const runIteration = () => {
+        // TODO: #9366746: integrate RelayRenderer/Container with GC hold
+        warning(
+          !pendingQueryTracker.hasPendingQueries(),
+          'RelayGarbageCollection: GC is executing during a fetch, but the ' +
+          'pending query may rely on data that is collected.'
+        );
+        let iterations = 0;
+        let hasNext = true;
+        while (hasNext && (stepLength < 0 || iterations < stepLength)) {
+          hasNext = run();
+          iterations++;
+        }
+        // This is effectively a (possibly async) `while` loop
+        if (hasNext) {
+          RelayTaskScheduler.enqueue(runIteration);
+        }
+      };
+      RelayTaskScheduler.enqueue(runIteration);
+    }
```

Comment:
Thanks for considering this, but let's move the scheduler back to `RelayGarbageCollection`